File: src/python/fbp/timeseries_to_data.py
#!/usr/bin/python
# -*- coding: UTF-8

# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/. */

# Authors:
# Michael Berg-Mohnicke <michael.berg@zalf.de>
#
# Maintainers:
# Currently maintained by the authors.
#
# This file is part of the util library used by models created at the Institute of
# Landscape Systems Analysis at the ZALF.
# Copyright (C: Leibniz Centre for Agricultural Landscape Research (ZALF)

import capnp
import io
import os
from pathlib import Path
from pyproj import CRS, Transformer
import sys
from datetime import timedelta, date

PATH_TO_SCRIPT_DIR = Path(os.path.realpath(__file__)).parent
PATH_TO_REPO = PATH_TO_SCRIPT_DIR.parent.parent.parent
if str(PATH_TO_REPO) not in sys.path:
    sys.path.insert(1, str(PATH_TO_REPO))

PATH_TO_PYTHON_CODE = PATH_TO_REPO / "src/python"
if str(PATH_TO_PYTHON_CODE) not in sys.path:
    sys.path.insert(1, str(PATH_TO_PYTHON_CODE))

import common.common as common
import common.geo as geo

PATH_TO_CAPNP_SCHEMAS = PATH_TO_REPO / "capnproto_schemas"
abs_imports = [str(PATH_TO_CAPNP_SCHEMAS)]
common_capnp = capnp.load(str(PATH_TO_CAPNP_SCHEMAS / "common.capnp"), imports=abs_imports) 
geo_capnp = capnp.load(str(PATH_TO_CAPNP_SCHEMAS / "geo.capnp"), imports=abs_imports)
climate_capnp = capnp.load(str(PATH_TO_CAPNP_SCHEMAS / "climate.capnp"), imports=abs_imports)

#------------------------------------------------------------------------------

config = {
    "in_sr": None, # string (sturdyref) | climate_capnp.TimeSeries (capability)
    "in_type": "sturdyref", # sturdyref | capability 

    "out_sr": None, # climate_capnp.TimeSeriesData (data)

    "to_attr": None, #"latlon",
    "from_attr": None,

    "subrange_start": None, # iso-date string
    "subrange_end": None, # iso-date string
    "subheader": None, # precip,globrad,tavg   ... etc
    "transposed": "false",
}
common.update_config(config, sys.argv, print_config=True, allow_new_keys=False)

conman = common.ConnectionManager()
inp = conman.try_connect(config["in_sr"], cast_as=common_capnp.Channel.Reader, retry_secs=1)
outp = conman.try_connect(config["out_sr"], cast_as=common_capnp.Channel.Writer, retry_secs=1)

in_type = config["in_type"]

def create_capnp_date(isodate):
    py_date = date.fromisoformat(isodate)
    return {"year": py_date.year, "month": py_date.month, "day": py_date.day}

def set_capnp_date(capnp_date, isodate):
    py_date = date.fromisoformat(isodate)
    capnp_date.year = py_date.year
    capnp_date.month = py_date.month
    capnp_date.day = py_date.day

try:
    if inp and outp:
        while True:
            msg = inp.read().wait()
            # check for end of data from in port
            if msg.which() == "done":
                break
            
            in_ip = msg.value.as_struct(common_capnp.IP)
            attr = common.get_fbp_attr(in_ip, config["from_attr"])
            obj = attr if attr else in_ip.content
            if in_type == "capability":
                timeseries = obj.as_interface(climate_capnp.TimeSeries)
            elif in_type == "sturdyref":
                try:
                    timeseries = conman.try_connect(obj.as_text(), cast_as=climate_capnp.TimeSeries, retry_secs=1)
                except Exception as e:
                    print("Error: Couldn't connect to sturdyref:", obj.as_text())
                    continue

            tsd = climate_capnp.TimeSeriesData.new_message()
            tsd.isTransposed = config["transposed"] == "true"
            
            if config["subheader"]:
                subheader = config["subheader"].split(",")
                timeseries = timeseries.subheader(subheader).timeSeries
            header = timeseries.header().wait().header
            
            if config["subrange_start"] or config["subrange_to"]:
                sr_req = timeseries.subrange_request()
                #timeseries = timeseries.subrange(
                #    ({"from": create_capnp_date(config["subrange_start"])} if config["subrange_start"] else {}) | 
                #    ({"to": create_capnp_date(config["subrange_end"])} if config["subrange_end"] else {})).timeSeries               
                #timeseries = timeseries.subrange(create_capnp_date(config["subrange_start"]), create_capnp_date(config["subrange_end"])).timeSeries
                if config["subrange_start"]:
                    set_capnp_date(sr_req.start, config["subrange_start"])
                if config["subrange_end"]:
                    set_capnp_date(sr_req.end, config["subrange_end"])
                timeseries = sr_req.send().wait().timeSeries

            resolution_prom = timeseries.resolution()
            se_date_prom = timeseries.range()
            header_size = len(header)
            ds = timeseries.dataT().wait().data if tsd.isTransposed else timeseries.data().wait().data
            tsd.init("data", len(ds))
            for i in range(len(ds)):
                l = tsd.data.init(i, header_size)
                for j in range(header_size):
                    l[j] = ds[i][j]
            se_date = se_date_prom.wait()
            tsd.startDate = se_date.startDate
            tsd.endDate = se_date.endDate
            tsd.resolution = resolution_prom.wait().resolution
            h = tsd.init("header", len(header))
            for i in range(len(header)):
                h[i] = header[i]

            out_ip = common_capnp.IP.new_message()
            if not config["to_attr"]:
                out_ip.content = tsd
            common.copy_and_set_fbp_attrs(in_ip, out_ip, **({config["to_attr"]: tsd} if config["to_attr"] else {}))
            outp.write(value=out_ip).wait()

        # close out port
        outp.write(done=None).wait()

except Exception as e:
    print("timeseries_to_data.py ex:", e)

print("timeseries_to_data.py: exiting run")

File: src/cpp/fbp/timeseries-to-data.cpp
/* This Source Code Form is subject to the terms of the Mozilla Public
* License, v. 2.0. If a copy of the MPL was not distributed with this
* file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/*
Authors:
Michael Berg <michael.berg@zalf.de>

Maintainers:
Currently maintained by the authors.

This file is part of the MONICA model.
Copyright (C) Leibniz Centre for Agricultural Landscape Research (ZALF)
*/

#include <iostream>
#include <fstream>
#include <string>
#include <tuple>
#include <vector>
#include <algorithm>

#include <kj/debug.h>
#include <kj/common.h>
#include <kj/exception.h>
#define KJ_MVCAP(var) var = kj::mv(var)
#include <kj/main.h>

#include <capnp/any.h>

#include "tools/date.h"
#include "tools/helper.h"
#include "common/rpc-connection-manager.h"
#include "common/common.h"

#include "climate.capnp.h"
#include "common.capnp.h"

using namespace std;
using namespace Tools;

namespace mas {
namespace infrastructure {
namespace fbp {

class TimeSeriesToTimeSeriesDataMain {
public:
  TimeSeriesToTimeSeriesDataMain(kj::ProcessContext &context) : context(context) {}

  kj::MainBuilder::Validity setName(kj::StringPtr n) { name = str(n); return true; }

  kj::MainBuilder::Validity setFromAttr(kj::StringPtr name) { fromAttr = kj::str(name); return true; }

  kj::MainBuilder::Validity setToAttr(kj::StringPtr name) { toAttr = kj::str(name); return true; }

  kj::MainBuilder::Validity setInSr(kj::StringPtr name) { inSr = kj::str(name); return true; }

  kj::MainBuilder::Validity setInType(kj::StringPtr type) { inType = kj::str(type); return true; }

  kj::MainBuilder::Validity setSubrangeStart(kj::StringPtr isoDateStr) { 
    subrangeStart = Tools::Date::fromIsoDateString(isoDateStr.cStr());
    return true; 
  }
  
  kj::MainBuilder::Validity setSubrangeEnd(kj::StringPtr isoDateStr) { 
    subrangeEnd = Tools::Date::fromIsoDateString(isoDateStr.cStr()); 
    return true; 
  }
  
  kj::MainBuilder::Validity setSubheader(kj::StringPtr headers) { 
    for(auto& s : mas::infrastructure::common::splitString(headers, ",")) subheaders.add(kj::str(s)); 
    return true; 
  }
  
  kj::MainBuilder::Validity setTransposed(kj::StringPtr isTransposed) { 
    transposed = Tools::toLower(isTransposed.cStr()) == "true"; 
    return true; 
  }

  kj::MainBuilder::Validity setOutSr(kj::StringPtr name) { outSr = kj::str(name); return true; }

  kj::MainBuilder::Validity startComponent() {
    auto ioContext = kj::setupAsyncIo();

    KJ_LOG(INFO, "starting timeseries-to-data Cap'n Proto FBP component");

    typedef schema::common::IP IP;
    typedef schema::common::Channel<IP> Channel;

    auto inp = conMan.tryConnectB(ioContext, inSr.cStr()).castAs<Channel::ChanReader>();
    auto outp = conMan.tryConnectB(ioContext, outSr.cStr()).castAs<Channel::ChanWriter>();

    using TS = mas::schema::climate::TimeSeries;
    using TSD = mas::schema::climate::TimeSeriesData;

    try {
      while(true) {
        auto msg = inp.readRequest().send().wait(ioContext.waitScope);
        // check for end of data from in port
        if (msg.isDone()) break;
        else {
          auto inIp = msg.getValue();
          auto attr = common::getIPAttr(inIp, fromAttr);
          TS::Client timeseries(nullptr);
          if (inType == "capability") {
            timeseries = attr.orDefault(inIp.getContent()).getAs<TS>();
          } else if(inType == "sturdyref") {
            auto timeseriesSR = attr.orDefault(inIp.getContent()).getAs<capnp::Text>();
            timeseries = conMan.tryConnectB(ioContext, timeseriesSR).castAs<TS>();
          }

          auto wreq = outp.writeRequest();
          IP::Builder outIp = wreq.initValue();

          TSD::Builder tsdb(nullptr);
          // set content if not to be set as attribute
          if (kj::size(toAttr) == 0) tsdb = outIp.initContent().getAs<TSD>();
          // copy attributes, if any and set result as attribute, if requested
          KJ_IF_MAYBE(builder, common::copyAndSetIPAttrs(inIp, outIp, toAttr)) { 
            tsdb = builder->getAs<TSD>();
          }

          tsdb.setIsTransposed(transposed);
          if(subheaders.size() > 0) {
            auto shReq = timeseries.subheaderRequest();
            auto shb = shReq.initElements(subheaders.size());
            using E = mas::schema::climate::Element;
            for(size_t i = 0; i < subheaders.size(); ++i) {
              if(subheaders[i] == "tmin") shb.set(i, E::TMIN); 
              else if(subheaders[i] == "tavg") shb.set(i, E::TAVG); 
              else if(subheaders[i] == "tmax") shb.set(i, E::TMAX); 
              else if(subheaders[i] == "precip") shb.set(i, E::PRECIP); 
              else if(subheaders[i] == "relhumid") shb.set(i, E::RELHUMID); 
              else if(subheaders[i] == "wind") shb.set(i, E::WIND); 
              else if(subheaders[i] == "globrad") shb.set(i, E::GLOBRAD); 
            }
            timeseries = shReq.send().getTimeSeries();   
          }

          if(subrangeStart.isValid() || subrangeEnd.isValid()) {
            auto srReq = timeseries.subrangeRequest();
            if(subrangeStart.isValid()){
              auto start = srReq.initStart();
              start.setYear(subrangeStart.year());
              start.setMonth(subrangeStart.month());
              start.setDay(subrangeStart.day());
            }
            if(subrangeEnd.isValid()){
              auto end = srReq.initEnd();
              end.setYear(subrangeEnd.year());
              end.setMonth(subrangeEnd.month());
              end.setDay(subrangeEnd.day());
            }
            timeseries = srReq.send().getTimeSeries();   
          }

          auto resolutionPromise = timeseries.resolutionRequest().send();
          auto headerPromise = timeseries.headerRequest().send();
          auto rangePromise = timeseries.rangeRequest().send();
          if(transposed) tsdb.setData(timeseries.dataTRequest().send().wait(ioContext.waitScope).getData());
          else tsdb.setData(timeseries.dataRequest().send().wait(ioContext.waitScope).getData());

          tsdb.setResolution(resolutionPromise.wait(ioContext.waitScope).getResolution());
          tsdb.setHeader(headerPromise.wait(ioContext.waitScope).getHeader());
          auto range = rangePromise.wait(ioContext.waitScope);
          tsdb.setStartDate(range.getStartDate());
          tsdb.setEndDate(range.getEndDate());

          wreq.send().wait(ioContext.waitScope);
        }
      }

      auto wreq = outp.writeRequest();
      wreq.setDone();
      wreq.send().wait(ioContext.waitScope);
    } catch(kj::Exception e) {
      std::cerr << "Exception: " << e.getDescription().cStr() << endl;
    }

    return true;
  }

  kj::MainFunc getMain() {
    return kj::MainBuilder(context, "timeseries-to-data FBP Component v0.1", "Transform a timeseries cap into a pure data.")
      .addOptionWithArg({'n', "name"}, KJ_BIND_METHOD(*this, setName),
                        "<component-name>", "Give this component a name.")
      .addOptionWithArg({'f', "from_attr"}, KJ_BIND_METHOD(*this, setFromAttr),
                        "<attr>", "Which attribute to read the MONICA env from.")
      .addOptionWithArg({'t', "to_attr"}, KJ_BIND_METHOD(*this, setToAttr),
                        "<attr>", "Which attribute to write the MONICA result to.")
      .addOptionWithArg({'i', "in_sr"}, KJ_BIND_METHOD(*this, setInSr),
                        "<sturdy_ref>", "Sturdy ref to input channel.")
      .addOptionWithArg({"in_type"}, KJ_BIND_METHOD(*this, setInType),
                         "<sturdyref | capability (default: sturdyref)>", "Receive on in port a TimeSeries cap or a sturdy ref to one.")      
      .addOptionWithArg({'o', "out_sr"}, KJ_BIND_METHOD(*this, setOutSr),
                        "<sturdy_ref>", "Sturdy ref to output channel.")
      .addOptionWithArg({"subrange_start"}, KJ_BIND_METHOD(*this, setSubrangeStart),
                        "<iso-date-string>", "Start of subrange.")
      .addOptionWithArg({"subrange_end"}, KJ_BIND_METHOD(*this, setSubrangeEnd),
                        "<iso-date-string>", "End of subrange.")
      .addOptionWithArg({"subheader"}, KJ_BIND_METHOD(*this, setSubheader),
                        "<[precip,globrad,tavg,....] (default: all available)>", "List of climate elements.")
      .addOptionWithArg({"transposed"}, KJ_BIND_METHOD(*this, setTransposed),
                        "<true | false (default: false>", "Return data transposed.")
      .callAfterParsing(KJ_BIND_METHOD(*this, startComponent))
      .build();
  }

private:
  mas::infrastructure::common::ConnectionManager conMan;
  kj::String name;
  kj::ProcessContext &context;
  kj::String inSr;
  kj::String inType{kj::str("sturdyref")};
  kj::String outSr;
  kj::String fromAttr;
  kj::String toAttr;
  Tools::Date subrangeStart;
  Tools::Date subrangeEnd;
  kj::Vector<kj::String> subheaders;
  bool transposed{false};
};

} // namespace fbp
} // namespace infrastructure
} // namespace mas

KJ_MAIN(mas::infrastructure::fbp::TimeSeriesToTimeSeriesDataMain)

